### PR TITLE
Implement entity listener priority

### LINKED ISSUE #154: Arbitrary order for EntityListener
The order in which the EntityListeners are executed is arbitrary (entries of an ObjectMap).
Let's take an example.
I have two components, A and B.
I also have two EntityListeners. One that tracks changes in Familly.all(A.class) and the other that tracks changes in Familly.all(B.class).
When removing an entity that contains a Component A and a Component B from the engine, both listeners are called. The order in which they are called is arbitrary (it is basically the order in which the Engine got to know about the Families).

I have many cases where it is really important that one entity listener is executed before the other. 
Long story short, there should be a notion of priority for entity listeners.

### PR DESCRIPTION
**Pull request by Antag99**
This pull request implements entity listener priority. Fixes #154.

### CODE REVIEW

**Comment by Antag99** [comment]
Closing this, I'll reopen it when I'm done.

**Comment by saltares** [comment]
Awesome.

Once you're done, do you mind squashing all commits into one?

**Comment by Antag99** [comment]
No, I don't mind, but is it possible to do it without breaking this pull request?

**Comment by saltares** [comment]
You will need to push to your branch from which you started this PR using `--force`.

**Comment by Antag99** [comment]
Well, I am done now, but I haven't gotten my pull request for libgdx merged yet, which fixes some bugs in the `Bits` class that affect my implementation. Any suggestions for additional unit tests?

**Comment by Antag99** [comment]
@saltares, could you update the libgdx version to `1.6.2`?

**Comment by Antag99** [comment]
Never mind, I updated it and did a rebase.

**Comment by Antag99** [comment]
@saltares Do you have time to review this?

**Comment by saltares** [comment]
I'll try to review this as soon as possible, sorry for the delay.

File: ashley/src/com/badlogic/ashley/core/Engine.java [hunk 1]
```diff
@@ -220,8 +222,17 @@ public void removeSystem(EntitySystem system){
...
+	public void addEntityListener (EntityListener listener) {
+		addEntityListener(empty, 0, listener);
+	}
+
+	/**
+	 * Adds an {@link EntityListener}. The listener will be notified every time an entity is added/removed
+	 * to/from the engine. The priority determines in which order the entity listeners will be called. Lower
+	 * value means it will get executed first.
+	 */
+	public void addEntityListener (int priority, EntityListener listener) {
```
Comment (saltares) [inline ashley/src/com/badlogic/ashley/core/Engine.java]: I'd prefer:

`addEntityListener (EntityListener listener, int priority)`

File: ashley/src/com/badlogic/ashley/core/Engine.java [hunk 2]
```diff
@@ -230,25 +241,66 @@ public void addEntityListener(EntityListener listener) {
...
 	public void addEntityListener(Family family, EntityListener listener) {
+		addEntityListener(family, 0, listener);
+	}
+
+	/**
+	 * Adds an {@link EntityListener} for a specific {@link Family}. The listener will be notified every time an entity is
+	 * added/removed to/from the given family. The priority determines in which order the entity listeners will be called. Lower
+	 * value means it will get executed first.
+	 */
+	public void addEntityListener (Family family, int priority, EntityListener listener) {
```
Comment (saltares) [inline ashley/src/com/badlogic/ashley/core/Engine.java]: Wouldn't it be nice to do?

`addEntityListener (EntityListener listener, Family family, int priority)`

**Comment by saltares** [comment]
I'm not too sure about adding Mockito as a dependency just for 2 tests. Surely we can make more tests use the library but maybe that'd belong into another changeset.

To be honest, I'm not familiar with Mockito in particular, although I've used other mocking frameworks in the past. What do you think of it?

File: ashley/src/com/badlogic/ashley/core/Engine.java (hunk 1 above)
Comment (Antag99) [inline ashley/src/com/badlogic/ashley/core/Engine.java]: Are you really sure about that? This seems fancier to me:
```
engine.addEntityListener(2, new EntityListener() {
    ...
});
```

**Comment by Antag99** [comment]
Well, I've used it for these two tests, but so far it's awesome. I just couldn't find an elegant way to test the invocation order.

**Comment by saltares** [comment]
Yeah, invocation order sounds nasty to test in an elegant way. Anyway, this is just an added dependency for the tests part of the project and won't ship to production code so I guess it's fine.

It looks good to me, can you merge from master so I can accept it?

Thanks!

**Comment by Antag99** [comment]
Done. I've resolved the merge conflicts and rebased it.

**Comment by Antag99** [comment]
I added some calls to `verifyNoMoreInteractions()`, though it's not likely to catch additional bugs.

**Comment by saltares** [comment]
I had to merge it manually but finally it's in! Thanks a lot.

**Comment by saltares** [comment]
Also, updated the wiki https://github.com/libgdx/ashley/wiki/How-to-use-Ashley#entity-events

**Pull request merged**